### PR TITLE
Nuke slice_as{,_mut}_ptr methods of MaybeUninit

### PR DESCRIPTION
**Pull request by SUPERCILEX**
These methods are of questionable value: they mix together removing bounds checks with converting the MaybeUninit slice into a raw pointer. These two behaviors should be explicitly invoked separately to improve code clarity and safety. There is evidence of unnecessary bounds checking removal in the stdlib uses that are updated in this PR.

In https://github.com/rust-lang/libs-team/issues/122, there is discussion of a need for generalized containers. However, that's going to be a long ways off, so in the meantime this PR adds an `into_inner` method on MaybeUninit raw pointers to unwrap the raw pointer. These aren't strictly necessary and could be replaced by `cast`, but then you don't get and type system guarentees.

### CODE REVIEW

File: library/core/src/fmt/num.rs [hunk 1]
```diff
@@ -480,7 +476,7 @@ impl_Exp!(i128, u128 as u128 via to_u128 named exp_u128);
 
 /// Helper function for writing a u64 into `buf` going from last to first, with `curr`.
 fn parse_u64_into<const N: usize>(mut n: u64, buf: &mut [MaybeUninit<u8>; N], curr: &mut usize) {
-    let buf_ptr = MaybeUninit::slice_as_mut_ptr(buf);
+    let buf_ptr = buf.as_mut_ptr().inner();
```
Comment (SUPERCILEX) [inline library/core/src/fmt/num.rs:479]: This could potentially be replaced with a bunch of `MaybeUninit::write`s, but there are so many of them that I don't want to risk bounds checking perf.

Comment (SUPERCILEX) [reply library/core/src/fmt/num.rs:479]: `add` is already unsafe, so I guess probs not needed.

**Comment by RalfJung** [comment]
I'm not a libs API person, so I'm the wrong reviewer here.
r? libs

File: compiler/rustc_serialize/src/opaque.rs [hunk 2]
```diff
@@ -346,7 +341,7 @@ impl FileEncoder {
             // room to write the input to the buffer.
             unsafe {
                 let src = buf.as_ptr();
-                let dst = MaybeUninit::slice_as_mut_ptr(&mut self.buf).add(buffered);
+                let dst = self.buf[buffered].as_mut_ptr();
```
Comment (RalfJung) [inline compiler/rustc_serialize/src/opaque.rs:344]: This adds a runtime bounds check where there was none before. Same for all the other places where `add` became `[...]`.

**Comment by SUPERCILEX** [comment]
Right, keep forgetting, sorry! :)

File: compiler/rustc_serialize/src/opaque.rs (hunk 2 above)
Comment (SUPERCILEX) [reply compiler/rustc_serialize/src/opaque.rs:344]: Correct. But the unsafety is centered around `copy_nonoverlapping` and presumably llvm can optimize out the bounds checking since it's done right above?

Happy to remove the bounds checking if that's what reviewers want, but either way `get_unchecked` is the method designed to express the desire to avoid bounds checking, not pointer addition IMO.

Comment (RalfJung) [reply compiler/rustc_serialize/src/opaque.rs:344]: I don't know if the bounds checks here are a problem, I only know I took care to not introduce any new bounds checks when this code was originally ported to `MaybeUninit`.

File: library/core/src/mem/maybe_uninit.rs [hunk 3]
```diff
@@ -1321,3 +1305,29 @@ impl<T, const N: usize> [MaybeUninit<T>; N] {
         unsafe { super::transmute_copy(&ManuallyDrop::new(self)) }
     }
 }
+
+// --- FIXME ---
+// Stabilize these as From implementations.
+// The only reason they aren't From is because that's insta-stable.
```
Comment (RalfJung) [inline library/core/src/mem/maybe_uninit.rs:1311]: IMO `into` is *way* to implicit for use around unsafe code. Unsafe code needs clearly stated intent, so just saying "please convert this to something else" is not explicit enough.

So IMO we actually want dedicated methods here, and we need to bikeshed their name. I don't like `inner` but I also don't have better suggestions at the moment.

**Comment by Nilstrieb** [comment]
If you have to add a completely new method to actually replace these methods, then maybe they aren't ready for removal imo. You should add the new method through the correct processes first, and then we can maybe remove these later.

Also I agree with Ralf that these should not be From impls at all.

**Comment by SUPERCILEX** [comment]
Agreed. See https://github.com/rust-lang/libs-team/issues/122#issuecomment-1281408047. Let's let this sit until the other MaybeUninit stuff is resolved and I'll give this some more thought.

@rustbot author

**Comment by JohnCSimon** [comment]
@SUPERCILEX 
ping from triage - can you post your status on this PR? There hasn't been an update in a few months. Is thanks still blocked? Thanks!

File: library/core/src/mem/maybe_uninit.rs (hunk 3 above)
Comment (SUPERCILEX) [reply library/core/src/mem/maybe_uninit.rs]: I went with `into_inner` to be consistent with other APIs.

File: compiler/rustc_serialize/src/opaque.rs (hunk 2 above)
Comment (SUPERCILEX) [reply compiler/rustc_serialize/src/opaque.rs]: I went back through and removed all the bounds checks that wouldn't be optimized out by the compiler. Anything remaining should be trivially removable.

**Comment by SUPERCILEX** [comment]
I don't think we had a clear resolution on this method specifically, but I stand by my original statement that these methods do too much at once. Let's throw this back in the hat @rustbot ready

**Comment by SUPERCILEX** [comment]
cc @m-ou-se

**Comment by SUPERCILEX** [comment]
> If you have to add a completely new method to actually replace these methods, then maybe they aren't ready for removal imo.

Regarding the objection that these are just replacement methods, I don't think I agree. The `into_inner` methods are more general: they let you safely get the underlying raw pointer of MaybeUninit pointer. Ideally this won't be needed if we get generalized containers, but in the meantime this is useful beyond just working with arrays.

**Comment by SUPERCILEX** [comment]
I just realized that some of this requires tree borrows to be nice. Added a commit that can be reverted once those become the default.

File: library/core/src/mem/maybe_uninit.rs (hunk 3 above)
Comment (RalfJung) [reply library/core/src/mem/maybe_uninit.rs]: I don't know any API this is consistent with? `into_inner` is usually for methods of the form `Type<T> -> T`; I find it extremely surprising to see it here for a raw pointer method.

Something like `as_inner_ptr` or so maybe?

Comment (SUPERCILEX) [reply library/core/src/mem/maybe_uninit.rs]: I guess I don't see `*const Type<T> -> *const T` as being too different from `Type<T> -> T`. I guess the difference is that one is a type cast whereas the other is providing a view into a type.

`as_` is supposed to be for borrowed types—are we counting pointers as a kind of borrow for naming purposes? If so `as_inner_ptr` seems great or maybe even just `as_inner`.

Comment (RalfJung) [reply library/core/src/mem/maybe_uninit.rs]: > I don't see *const Type<T> -> *const T as being too different from Type<T> -> T

Why not? It makes a huge difference whether ownership is consume vs whether there is a form of "borrowing". We never use `into_inner` for methods like `&Type<T> -> &T` or `&mut Type<T> -> &mut T`. These functions here are very much like them, except they are borrowing via raw pointers, not references.

One similar function we have is [`UnsafeCell::raw_get`](https://doc.rust-lang.org/nightly/core/cell/struct.UnsafeCell.html#method.raw_get). On `MaybeUninit` we already have `as_(mut_)ptr` that take references; now here we have the exact same methods using raw pointers so the naming should be similar. `raw_as_ptr` sounds odd though...

**Comment by Mark-Simulacrum** [comment]
@rustbot author

I think the new API surface area does need a signed-off ACP, so moving this back to waiting on author. The code changes don't feel like a win to me (or a huge loss, either), so I don't know that this is that much clearer. I share Ralf's opinion that into_inner does not strike me as a great name for `*const/*mut MaybeUninit<T> -> *const/*mut T`, and that in most cases in this PR `.cast::<u8>()` or similar would be perfectly fine.

File: library/core/src/mem/maybe_uninit.rs (hunk 3 above)
Comment (SUPERCILEX) [reply library/core/src/mem/maybe_uninit.rs]: Gotya, I wasn't really thinking of pointers as references for naming but you're totally right that makes more sense. `raw_as_ptr` does seems a little weird, but it would be consistent... will rename.

**Comment by SUPERCILEX** [comment]
> in most cases in this PR .cast::<u8>() or similar would be perfectly fine.

Do you think we just shouldn't bother adding a replacement method? I would like to see the slice ptr methods removed, but don't think it's absolutely necessary to have the type safety (though it'd be a slight downgrade).

> I think the new API surface area does need a signed-off ACP

Should I make a new ACP for this? The existing one seems pretty stalled.

**Comment by Mark-Simulacrum** [comment]
I think the libs-api team is continuing to work through the ACP backlog, but it seems like the existing one is somewhat indirectly referencing the specific proposal made in this PR, which may make sense to do independently.

**Comment by RalfJung** [comment]
You mean https://github.com/rust-lang/libs-team/issues/122? That covers a lot of things, the methods added here are just a tiny part of it. (Might be good to update the ACP for the new method names.)

The raw versions of `as_ptr`/`as_mut_mtr` do seem like a natural addition to the API surface to me. Not sure how the libs team prefers to stage this -- you might ask on Zulip if they prefer a small ACP for just these 2 methods, vs a big ACP for the entire revamp.

**Comment by SUPERCILEX** [comment]
Alrighty, I created https://github.com/rust-lang/libs-team/issues/245 because I think creating a small, focused ACP will help move things forward.

File: compiler/rustc_serialize/src/opaque.rs [hunk 4]
```diff
@@ -182,8 +182,8 @@ impl FileEncoder {
             // room to write the input to the buffer.
             unsafe {
                 let src = buf.as_ptr();
-                let dst = MaybeUninit::slice_as_mut_ptr(&mut self.buf).add(buffered);
-                ptr::copy_nonoverlapping(src, dst, buf_len);
+                let dst = self.buf.as_mut_ptr().add(buffered);
+                ptr::copy_nonoverlapping(src, dst.cast(), buf_len);
```
Comment (ChrisDenton) [inline compiler/rustc_serialize/src/opaque.rs:186]: Minor nit but could you make casts like this more explicit? When looking at unsafe code I prefer to have as few things to think about as possible.

Comment (SUPERCILEX) [reply compiler/rustc_serialize/src/opaque.rs:186]: As in `.cast::<u8>()` (and moved up to the `dst` line)? Happy to do that, though I'm curious as to why that's safer. Is it because we're assuming the cast is correct and relying on the type system to complain when pointer types aren't matched up later?

Comment (ChrisDenton) [reply compiler/rustc_serialize/src/opaque.rs:186]: Yeah. As I say, it's a minor thing here because the code is fairly straight forward. But in `unsafe` blocks I tend to prefer to keep things very obvious and have as many sanity checks as possible because the nature of it is that the compiler can't always check as much as in safe code. And future refactoring can have a nasty habit of changing things in unexpected ways so I like to make sure the humans and the compiler all clearly agree on what's happening even the human is just looking at a few lines of a diff.

In short, maybe I'm a bit paranoid but I think there's no harm in being too cautious.

Comment (SUPERCILEX) [reply compiler/rustc_serialize/src/opaque.rs:186]: No no, that makes sense, thanks!

**Comment by workingjubilee** [comment]
Do we need FCP or something for posterity, or should this just be r+'d? Can I even 
r? libs-api

**Comment by RalfJung** [comment]
I don't think I like this stripped down version - every new use of 'cast' is a step backwards IMO. These are hard to review since one cannot easily tell what they are being cast *from*.

**Comment by RalfJung** [comment]
> There is evidence of unnecessary bounds checking removal in the stdlib uses that are updated in this PR.

Which ones did you think were unnecessary? As far as I am aware, most/all of these are ported from old pre-MaybeUninit code that already did not have bounds checks. So the removal was entirely deliberate.

FWIW I agree that with a raw ptr projection method (`*mut MaybeUninit<T> -> *mut T`)  this becomes clearer, albeit more verbose. But reverting to `cast` strikes me as a bad intermediate step. It makes the code harder to read and it is unlikely to be ported to the raw ptr projection method once that exists (since how would we even find the places where it can be used).

**Comment by SUPERCILEX** [comment]
> Which ones did you think were unnecessary?

Take a look at all the places I used slicing operators in num.rs for this commit: [`f2e9b40` (#103133)](https://github.com/rust-lang/rust/pull/103133/commits/f2e9b40b8a2cca1c6a4c28eeac0eb5cfdfb1a3f3#diff-5c3682e8974cc4b780f07f1a1fd2fed4332e66d7645dc72fa7c29dbdd785ccda). All of those should have probably eliminatable bounds checks.

> But reverting to cast strikes me as a bad intermediate step.

Yeah, that's fair. So do those two changes need to happen together?

**Comment by RalfJung** [comment]
"should have probably" is fairly weak. This is all unsafe code because it is considered performance-sensitive. And this code is old, it used unchecked accesses already before it used `MaybeUninit`:

https://github.com/rust-lang/rust/blob/c11e514e9dfd44eebe435408b32de4193d87860c/src/libcore/fmt/num.rs

Notice all the `offset`. They were added in https://github.com/rust-lang/rust/commit/ebf9e1aaf63e741cec122f3cea1afa6d62b01350. 

So this is definitely not evidence that the current API leads to omitted bounds checks.

> So do those two changes need to happen together?

I would personally prefer that. I don't have the power to block changes in this part of the compiler though.

**Comment by SUPERCILEX** [comment]
> "should have probably"

Typo that autocorrect keeps trying to put in, sorry. Should have **provably**. As in if they don't get optimized out it's an llvm bug.

> I would personally prefer that.

Yeah I think that's reasonable unfortunately. I'll have time to look into this in a month or two again.

**Comment by apiraino** [comment]
Based on the [prev. comment](https://github.com/rust-lang/rust/pull/103133#issuecomment-1662917984) switching to waiting on author to incorporate changes. Feel free to request a review with `@rustbot ready`, thanks!

@rustbot author

**Comment by JohnCSimon** [comment]
@SUPERCILEX 
ping from triage - can you post your status on this PR? There hasn't been an update in a few months. Thanks!


FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

**Comment by SUPERCILEX** [comment]
This PR is waiting on me to figure out what to do with `*{const,mut} MaybeUninit<T>` to `*{const,mut} T` ptr casts. Hoping to have time to think about this in December/January.

**Comment by SUPERCILEX** [comment]
I'll close for now. I still think these methods should be removed, but sadly don't have the time to see that work through right now.
